---
title: "Simulate Gene Expression Evolution Across Tree"
author: "Erin Fry"
date: "2/13/2017"
output:
  html_document:
    toc: true
    self_contained: false
---

**This script simulates continuous character evolution on a phylogenetic Tree. Specifically, I will simulate gene expression evolution of 10,000 genes across the primate tree used in my ancestral gene expression reconstruction (AGER) project.**

```{r setup, echo=FALSE, warning=FALSE}
## set paths and load packages
path="/Users/lynchlab/Desktop/ErinFry/ReconAncNeoTranscriptomes/BrainConstitiutive/BTReconstruct/"
pathData=paste(path,"data/",sep="")
pathResults=paste(path,"resultsSim/",sep="")


library("ape")
library("ggplot2")
```

### Brownian motion phylogenetic simulators are obtained from Phytools at https://github.com/liamrevell/phytools/tree/master/R
```{r phytools}
## internal function does BM simulation
## written by Liam J. Revell 2011, 2013
simBM<-function(tree,a,mu,sig2,bounds,internal,nsim){
	if(bounds[2]<bounds[1]){
		warning("bounds[2] must be > bounds[1]. Simulating without bounds.")
		bounds<-c(-Inf,Inf)
	}
	if(bounds[1]==-Inf&&bounds[2]==Inf) no.bounds=TRUE
	else no.bounds=FALSE
	if(a<bounds[1]||a>bounds[2]){
		warning("a must be bounds[1]<a<bounds[2]. Setting a to midpoint of bounds.")
		a<-bounds[1]+(bounds[2]-bounds[1])/2
	}
	if(sig2<0){
		warning("sig2 must be > 0.  Setting sig2 to 1.0.")
		sig2=1.0
	}
	# function for reflection off bounds
	reflect<-function(yy,bounds){
		while(yy<bounds[1]||yy>bounds[2]){
			if(yy<bounds[1]) yy<-2*bounds[1]-yy
			if(yy>bounds[2]) yy<-2*bounds[2]-yy
		}
		return(yy)
	}
	# how many species?
	n<-length(tree$tip)
	# first simulate changes along each branch
	x<-matrix(data=rnorm(n=length(tree$edge.length)*nsim,mean=rep(mu*tree$edge.length,nsim),sd=rep(sqrt(sig2*tree$edge.length),nsim)),length(tree$edge.length),nsim)
	# now add them up
	y<-array(0,dim=c(nrow(tree$edge),ncol(tree$edge),nsim))
	for(i in 1:nrow(x)){
		if(tree$edge[i,1]==(n+1))
			y[i,1,]<-a
		else
			y[i,1,]<-y[match(tree$edge[i,1],tree$edge[,2]),2,]

		y[i,2,]<-y[i,1,]+x[i,]
		if(!no.bounds) y[i,2,]<-apply(as.matrix(y[i,2,]),1,function(yy) reflect(yy,bounds))
	}
	rm(x); x<-matrix(data=rbind(y[1,1,],as.matrix(y[,2,])),length(tree$edge.length)+1,nsim)
	rownames(x)<-c(n+1,tree$edge[,2])
	x<-as.matrix(x[as.character(1:(n+tree$Nnode)),])
	rownames(x)[1:n]<-tree$tip.label
	# return simulated data
	if(internal==TRUE)
		return(x[1:nrow(x),]) # include internal nodes
	else
		return(x[1:length(tree$tip.label),]) # tip nodes only
}


## internal function does BM simulation
## written by Liam J. Revell 2013
simOU<-function(tree,alpha,sig2,theta,a0,internal){
	tree<-reorder(tree,"cladewise")
	X<-matrix(0,nrow(tree$edge),ncol(tree$edge))
	root<-length(tree$tip.label)+1
	X[which(tree$edge[,1]==root),1]<-a0
	for(i in 1:nrow(X)){
		t<-tree$edge.length[i]
		s2<-sig2*(1-exp(-2*alpha*t))/(2*alpha)
		X[i,2]<-exp(-alpha*t)*X[i,1]+(1-exp(-alpha*t))*theta+rnorm(n=1,sd=sqrt(s2))
		ii<-which(tree$edge[,1]==tree$edge[i,2])
		if(length(ii)>0) X[ii,1]<-X[i,2]
	}
	x<-sapply(1:max(tree$edge),function(x,y,tree) y[which(tree$edge==x)[1]],y=X,tree=tree)
	x<-setNames(x,c(tree$tip.label,1:tree$Nnode+length(tree$tip.label)))
	if(internal==TRUE)
		return(x) # include internal nodes
	else
		return(x[1:length(tree$tip.label)]) # tip nodes only
}


# Simulates BM evolution more quickly.
# A trend can be simulated by mu!=0.
# mu=0 is standard BM; mu<0 downward trend; mu>0 upward trend.
# Bounds can be simulated by bounds=c(>-Inf,<Inf).
# OU can be simulated by alpha>0.
# Written by Liam J. Revell 2011, 2013, 2015

fastBM<-function(tree,a=0,mu=0,sig2=1,bounds=c(-Inf,Inf),internal=FALSE,nsim=1,...){
	# some minor error checking
	if(!inherits(tree,"phylo")) stop("tree should be object of class \"phylo\".")
	## check to see if alpha & theta
	if(hasArg(alpha)) alpha<-list(...)$alpha
	else alpha<-NULL
	if(hasArg(theta)) theta<-list(...)$theta
	else theta<-NULL
	if(!is.null(alpha)&&is.null(theta)){
		cat("Warning: alpha but not theta specified in OU model, setting theta to a.\n")
		theta<-a
	}
	## check for OU w. trend (not permitted)
	if(!is.null(alpha)&&mu!=0)
		cat("Warning: OU with a trend not permitted. Trend parameter will be ignored.\n")
	## check for OU w. bounds (not permitted)
	if(!is.null(alpha)&&(bounds[1]!=-Inf||bounds[2]!=Inf))
		cat("Warning: OU with bounds not permitted. Bounds will be ignored.\n")
	## if BM
	if(is.null(alpha)) x<-simBM(tree,a,mu,sig2,bounds,internal,nsim)
	else x<-if(nsim==1) simOU(tree,alpha,sig2,theta,a,internal) else replicate(nsim,simOU(tree,alpha,sig2,theta,a,internal))
	x
}



```


### Upload our tree and sqrt transformed expression data to simulate data similar to true data
```{r Simulate Data}
## set phylogenetic tree
setwd(pathData)
MyTree<-read.nexus("sim.tree")

TPM<-read.table("TPM_sqrt.txt",sep='\t')

## Visualize tree and internal node labels
plot(MyTree)
nodelabels()

## set seed
set.seed(6)


```

### Simulate 10,000 gene expression and evolution across the tree
```{r simulate}

## simulate 10000 genes with roots normally distributed by the mean and standard deviation of real data mean and standard deviation expression

## set the number of genes to simulate
nsimul=10000

## draw a large set of root expression values, some will be negative. These numbers are drawn based on the mean expression of genes in our dataset
a=rnorm(n=nsimul*10, mean=(median(apply(TPM[,-1],2,mean))), sd=(sd(apply(TPM[,-1],2,mean))))
## choose only positive root values
apos<-a[which(a>0)][1:nsimul]

## simulate gene expression evolution across the tree from a variety of root values
simulatevarietyofexp<-function(tree, rootvector, nsimul){
   simulationvar<-matrix(ncol=nsimul,nrow=(length(tree$edge.length)+1)) 
   for (i in 1:nsimul){
     simulationvar[,i]<-fastBM(tree,rootvector[i], bounds=c(0,Inf), internal=TRUE,sig2=abs(rnorm(1,mean=0.00001,sd=0.0001)))
  }
  return(simulationvar)
}

## Use the positive values based on our data to simulate expression data
simreflection<-simulatevarietyofexp(MyTree, rootvector = apos, nsimul = nsimul)

```

### Compare simulated data to real data
```{r compare}
## compare how similar my simulations meant to reflect the real data are to the real data
summary(apply(simreflection[,-1],2,sd))
summary(apply(TPM[,-1],2,sd))

summary(apply(simreflection[,-1],2,mean))
summary(apply(TPM[,-1],2,mean))


hist(log(apply(TPM[,-1],2,mean)), xlim=c(log(sqrt(2)),5), freq=F, ylim=c(0,1), xlab="log(Gene Mean Expression Value (sqrt(TPM)))", main="Expression in Real v. Simulated Data, blue=simulated")

hist(log(apply(simreflection[,-1],2,mean)), add=T, border="blue", freq=F)

```


## Save a .txt file of the simulated internal and extant nodes to test AGER
```{r save simulated values}

rownames(simreflection)<-c(as.character(TPM[,1]), "AncPrimate","AncOrangApe","AncApe","AncHominini","AncHomo","AncPan", "AncChimp","AncBonobo","AncGorilla","AncOrangutan")

## save all simulated data points
setwd(pathData)
write.table(simreflection, "17-03-17Simulation.txt", sep="\t")

## save file to run AGER pipeline, extant nodes only
write.table(simreflection[1:19,], "sqrt_tpm_sims.txt", sep="\t")



simreflection[1:19,1:2]
```

## Instead null simulate end nodes
```{r}
## simulate 10000 genes with roots normally distributed by the mean and standard deviation of real data mean and standard deviation expression

## set the number of genes to simulate
nsimul=10000

## draw a large set of average expression values, some will be negative. These numbers are drawn based on the mean expression of genes in our dataset
a=rnorm(n=nsimul*10, mean=(median(apply(TPM[,-1],2,mean))), sd=(sd(apply(TPM[,-1],2,mean))))
## choose only positive average values
apos<-a[which(a>0)][1:nsimul]
## same for standard deviation
s=rnorm(n=nsimul*10, mean=1, sd=1)
spos<-s[which(s>0)][1:nsimul]


## simulate gene expression variation across the tree
simulatevarietyofexp<-function(tree, average, standarddev, nsimul){
   simulationvar<-matrix(ncol=nsimul,nrow=(length(tree$tip.label))) 
   for (i in 1:nsimul){
     simulationvar[,i]<-rnorm(n = (length(tree$tip.label)), mean = average[i], sd = standarddev )
  }
  return(simulationvar)
}

## Use the positive values based on our data to simulate expression data
simreflection<-simulatevarietyofexp(MyTree, average = apos, standarddev=spos, nsimul = nsimul)
## convert all negatives to 0
simreflection[simreflection<0] <- 0

## compare how similar my simulations meant to reflect the real data are to the real data
summary(apply(simreflection[,-1],2,sd))
summary(apply(TPM[,-1],2,sd))

summary(apply(simreflection[,-1],2,mean))
summary(apply(TPM[,-1],2,mean))

rownames(simreflection)<-c(as.character(TPM[,1]))

## save all simulated data points
setwd(pathData)
write.table(simreflection, "17-03-27Simulation.txt", sep="\t")

## save file to run AGER pipeline, extant nodes only
write.table(simreflection, "sqrt_tpm_sims.txt", sep="\t")
```